### PR TITLE
Fix consent request for attributes that don't have values.

### PR DESCRIPTION
**Pull request by Wathsara**
#### Purpose

Previously consent has been asked for attributes that don't have values in the profile. With this fix the consent will be asked for the user attributes that have values.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java [hunk 1]
```diff
@@ -249,6 +251,30 @@ private void removeClaimsWithoutConsent(AuthenticationContext context, ConsentCl
         removeUserClaimsFromContext(context, claimsWithoutConsent, spStandardDialect);
     }
 
+    /**
+     * Filter out the requested claims with the user attributes.
+     * @param consentClaimsData Claim metadata.
```
Comment (mevan-karu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java:256]:
```suggestion
     * Filter out the requested claims with the user attributes.
     *
     * @param consentClaimsData Claim metadata.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java [hunk 2]
```diff
@@ -249,6 +251,30 @@ private void removeClaimsWithoutConsent(AuthenticationContext context, ConsentCl
...
+     * @param user              Authenticated user.
+     * @return                  Filtered claim metadata.
+     */
+    private ConsentClaimsData removeClaimsNotInUserAttribute(ConsentClaimsData consentClaimsData,
+                                                             AuthenticatedUser user) {
+
+        List<ClaimMetaData> requestedClaims = consentClaimsData.getRequestedClaims();
```
Comment (mevan-karu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java:263]: Can `consentClaimsData` be null? If so, add a null check.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java [hunk 3]
```diff
@@ -249,6 +251,30 @@ private void removeClaimsWithoutConsent(AuthenticationContext context, ConsentCl
...
+        List<ClaimMetaData> filteredRequestClaims = new ArrayList<>();
+        Map<ClaimMapping, String> userAttributes = user.getUserAttributes();
+        for (ClaimMetaData claimMetaData : requestedClaims) {
```
Comment (mevan-karu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java:266]: Can `userAttributes` or `requestedClaims` be null? If so, add a null check before `for` loops.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java [hunk 4]
```diff
@@ -249,6 +251,30 @@ private void removeClaimsWithoutConsent(AuthenticationContext context, ConsentCl
         removeUserClaimsFromContext(context, claimsWithoutConsent, spStandardDialect);
     }
 
+    /**
+     * Filter out the requested claims with the user attributes.
+     * @param consentClaimsData Claim metadata.
+     * @param user              Authenticated user.
+     * @return                  Filtered claim metadata.
+     */
+    private ConsentClaimsData removeClaimsNotInUserAttribute(ConsentClaimsData consentClaimsData,
```
Comment (mevan-karu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java:260]:
```suggestion
    private ConsentClaimsData removeClaimsNotInUserAttributes(ConsentClaimsData consentClaimsData,
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java [hunk 5]
```diff
@@ -249,6 +252,30 @@ private void removeClaimsWithoutConsent(AuthenticationContext context, ConsentCl
...
+     * Filter out the requested claims with the user attributes.
+     *
+     * @param requestedClaims List of requested claims metadata.
+     * @param userAttributes  Authenticated users' attributes.
+     * @return                Filtered claims with user attributes.
+     */
+    private List<ClaimMetaData> removeClaimsNotInUserAttributes(List<ClaimMetaData> requestedClaims,
+                                                                      Map<ClaimMapping, String> userAttributes) {
+
+        List<ClaimMetaData> filteredRequestClaims = new ArrayList<>();
```
Comment (mevan-karu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java:265]:
```suggestion
        List<ClaimMetaData> filteredRequestedClaims = new ArrayList<>();
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java [hunk 6]
```diff
@@ -249,6 +252,30 @@ private void removeClaimsWithoutConsent(AuthenticationContext context, ConsentCl
         removeUserClaimsFromContext(context, claimsWithoutConsent, spStandardDialect);
     }
 
+    /**
+     * Filter out the requested claims with the user attributes.
+     *
+     * @param requestedClaims List of requested claims metadata.
+     * @param userAttributes  Authenticated users' attributes.
+     * @return                Filtered claims with user attributes.
```
Comment (GANGANI) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java:260]: Check formatting

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java [hunk 7]
```diff
@@ -249,6 +252,30 @@ private void removeClaimsWithoutConsent(AuthenticationContext context, ConsentCl
...
+     */
+    private List<ClaimMetaData> removeClaimsNotInUserAttributes(List<ClaimMetaData> requestedClaims,
```
Comment (GANGANI) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java:262]:
```suggestion
    private List<ClaimMetaData> removeConsentRequestedNullUserAttributes(List<ClaimMetaData> requestedClaims,
```
WDYT?

**Review by mevan-karu — APPROVED** [review]

**Review by IsuraD — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java [hunk 8]
```diff
@@ -225,6 +230,33 @@ private void removeClaimsWithoutConsent(AuthenticationContext context, ConsentCl
...
+                                                                         Map <String, String> claimMappings) {
+
+        List<ClaimMetaData> filteredRequestedClaims = new ArrayList<>();
+        if (requestedClaims != null && userAttributes != null && claimMappings != null) {
+            for (ClaimMetaData claimMetaData : requestedClaims) {
+                for (Map.Entry<ClaimMapping, String> attribute : userAttributes.entrySet()) {
+                    if (claimMetaData.getClaimUri()
+                            .equals(claimMappings.get(attribute.getKey().getLocalClaim().getClaimUri()))) {
+                        filteredRequestedClaims.add(claimMetaData);
+                        break;
```
Comment (Thumimku) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java:252]: Why we break with first hitpoit?

**Pull request merged**